### DISCUSSION TITLE
Issue with results from all Entity Types with MS Search data source

### DISCUSSION THREAD

**Discussion by mw-new-test** [microsoft-search/pnp-modern-search#2806] [Q&A]
**Version used**
4.8

**Describe the bug**
We are building an Enterprise Search site using PnP Modern Search. Our requirement is to replicate the behaviour of the "All" vertical available at site/tenant level via the MS Search on PnP Modern Search.
The default "All" vertical on SharePoint/Office/Bing shows results from SharePoint, People, Teams, Outlook, Drive and also Graph connectors within Result clusters.
On PnP Modern Search, we are not able to combine all entity types to allow users to search for SharePoint, People, Drive items, Confluence/Jira graph connectors in one results webpart.
Error: SearchRequest Invalid (EntityRequest Invalid (Invalid entity type combination)) Invalid entity type combination

**To Reproduce**
Detailed steps to reproduce the behavior:
1. Add verticals webpart named "All"
2. Add Search results webpart
3. Select Microsoft Search as Data Source
4. Select ListItem, Drive Item, ExternalItem, People, Sites etc. in Entity Type
5. Provide Content sources ID for External Item (Confluence, ServiceNow etc.)
6. Enable result types to apply a custom layout for ExternalItem
7. Connect Results webpart to vertical webpart
8. Error: SearchRequest Invalid (EntityRequest Invalid (Invalid entity type combination)) Invalid entity type combination
9. If we remove People from the List of Entity types, it shows some results but does not include results from externalItem content source

**Expected behavior**
Is the behaviour similar to MS Search "All" vertical possible with PnP Modern search?
Users should be able to see All results collated within one vertical and results from Graph connectors shown in Result clusters with the result type/custom layout applied.

**Desktop (please complete the following information):**
Chrome, Edge

**Additional context**

**Comment by wobba** [comment]
You have to follow what is possible in the Graph API in terms of joining multiple entity types. If you have more verticals then you can have different entities per vertical.

**Reply by mw-new-test** [reply]
Thank you for the response. 
I can have different/separate verticals for different entity types. 
But, can all entity types not be combined on PnP Modern Search the way it is done on Microsoft Search while maintaining result type for each of the ExternalItem results shown in the combined vertical?

**Reply by wobba** [reply]
The web parts depend on what is possible using the Graph Search API. If it works in the Graph Explorer it should work in the web parts.

**Reply by mw-new-test** [reply]
Excluding message, chatMessage, event and person, all other result entity types can be interleaved as per - https://learn.microsoft.com/en-us/graph/search-concept-interleaving
But Result Template can only be applied to ListItem and ExternalItem.

When we are combining ListItem with ExternalItem in PnP Modern Search Results webpart, it is currently only showing results for ListItem.
If we search for any keyword that should be part of ExternalItem content, it is returning the following error - Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')

**Reply by wobba** [reply]
Try to repro in the Graph Explorer. If it fails there, then it's an API issue. If not it could be an API crafting issue with the web parts.

**Reply by mw-new-test** [reply]
Thanks, I tried to replicate in Graph explorer with below Request and Body/Payload
POST - https://graph.microsoft.com/v1.0/search/query

{
    "requests": [
        {
            "entityTypes": [
                "externalItem",
                "listItem",
                "driveItem",
                "site",
                "drive",
                "list"
            ],
            "contentSources": [
                "/external/connections/*"
            ],
            "query": {
                "queryString": "lumix"
            },
            "from": 0,
            "size": 500
        }
    ]
}

I received 12 results including 3 ExternalItem results in Graph explorer.
However, in PnP Modern Search, it only returned 9 results without the 3 ExternalItem results.
The Entity types payload and Content sources in PnP Modern Search were same as the ones passed in Graph explorer.

Could you please check on your end and let me know if there is an issue with the API not returning ExternalItem results in PnP Modern Search Results webpart?

**Reply by wobba** [reply]
Can you compare the request JSON from the network tab when using PnP as it should be the same? I checked all six sources and had no issue.

![image](https://user-images.githubusercontent.com/51104/224724109-5e7982ee-9983-49f8-9314-9db739b72133.png)

Also used `*` as the content source as your sample.
![image](https://user-images.githubusercontent.com/51104/224724284-c52fe654-3631-4828-962f-b5e0db70f186.png)

**Reply by mw-new-test** [reply]
Thanks very much for checking. 
I am able to see results for ExternalItem now with "*" as the Content Source.
"/external/connections/*" what we use in Graph API/explorer, doesn't work in PnP Modern Search.
Glad that it worked with "*" .Thanks again for the help :)
Although now there is a different issue - for ExternalItem, it does not return any properties in PnP Modern Search.
On Graph Explorer, I can see properties (schema) of ExternalItem in the results.
But in PnP resource.properties is empty even in the network tab on developer tools. I can see properties only if i am providing the actual ID of the Content Source (like Confluence/Salesforce) but doesn't return properties with "*" Content source in PnP Modern Search webpart.
What could be the issue?

**Reply by wobba** [reply]
You have to either include a connection name or `*` to include all connections.
